packages/app/src/hooks/useChartConfig.tsx

```diff
@@ -1,57 +1,257 @@
+  /**
+   * By default, queries with large date ranges are split into multiple smaller queries to
+   * avoid overloading the ClickHouse server and running into timeouts. In some cases, such
+   * as when data is being sampled across the entire range, this chunking is not desirable
+   * and can be disabled.
+   */
+  disableQueryChunking?: boolean;
 }
 
-// used for charting
+type TimeWindow = {
+  dateRange: [Date, Date];
+  dateRangeEndInclusive?: boolean;
+};
+
+type TQueryFnData = Pick<ResponseJSON<any>, 'data' | 'meta' | 'rows'> & {
+  isComplete: boolean;
+};
+
+type TChunk = {
+  chunk: ResponseJSON<Record<string, string | number>>;
+  isComplete: boolean;
+};
+
+const shouldUseChunking = (
+  config: ChartConfigWithOptDateRange,
+): config is ChartConfigWithDateRange & {
+  granularity: string;
+} => {
+  // Granularity is required for chunking, otherwise we could break other group-bys.
+  if (!isUsingGranularity(config)) return false;
+
+  // Date range is required for chunking, otherwise we'd have infinite chunks, or some unbounded chunk(s).
+  if (!config.dateRange) return false;
+
+  // TODO: enable chunking for metric charts when we're confident chunking will not break
+  // complex metric queries.
+  if (isMetricChartConfig(config)) return false;
+
+  return true;
+};
+
+export const getGranularityAlignedTimeWindows = (
+  config: ChartConfigWithDateRange & { granularity: string },
+  windowDurationsSeconds?: number[],
+): TimeWindow[] => {
+  const [startDate, endDate] = config.dateRange;
+  const windowsUnaligned = generateTimeWindowsDescending(
+    startDate,
+    endDate,
+    windowDurationsSeconds,
+  );
+
+  const granularity =
+    config.granularity === 'auto'
+      ? convertDateRangeToGranularityString(
+          config.dateRange,
+          DEFAULT_AUTO_GRANULARITY_MAX_BUCKETS,
+        )
+      : config.granularity;
+
+  const windows = [];
+  for (const [index, window] of windowsUnaligned.entries()) {
+    // Align windows to chart buckets
+    const alignedStart =
+      index === windowsUnaligned.length - 1
+        ? window.startTime
+        : toStartOfInterval(window.startTime, granularity);
+    const alignedEnd =
+      index === 0 ? endDate : toStartOfInterval(window.endTime, granularity);
+
+    // Skip windows that are covered by the previous window after it was aligned
+    if (
+      !windows.length ||
+      alignedStart < windows[windows.length - 1].dateRange[0]
+    ) {
+      windows.push({
+        dateRange: [alignedStart, alignedEnd] as [Date, Date],
+        // Ensure that windows don't overlap by making all but the first (most recent) exclusive
+        dateRangeEndInclusive:
+          index === 0 ? config.dateRangeEndInclusive : false,
+      });
+    }
+  }
+
+  return windows;
+};
+
+async function* fetchDataInChunks(
+  config: ChartConfigWithOptDateRange,
+  clickhouseClient: ClickhouseClient,
+  signal: AbortSignal,
+  disableQueryChunking: boolean = false,
+) {
```

Comment:
style:I suggest converting it into an object when the method takes more than three args

Reply:
Good idea, done